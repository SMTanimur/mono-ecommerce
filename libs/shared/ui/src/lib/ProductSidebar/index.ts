import { ProductSidebar, ProductSidebarProps } from './ProductSidebar';

export { ProductSidebar };
export type { ProductSidebarProps };
