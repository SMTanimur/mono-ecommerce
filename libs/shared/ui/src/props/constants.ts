import { ButtonProps } from '@ecommerce/shared/ui/Button';
import {CartItemProps} from "@ecommerce/shared/ui/CartItem";
import {ProductSidebarProps} from "@ecommerce/shared/ui/ProductSidebar";
import {CartSidebarProps} from "@ecommerce/shared/ui/CartSidebar";
import {CategoryCardProps} from "@ecommerce/shared/ui/CategoryCard";
import {IconProps} from "@ecommerce/shared/ui/Icon";
import {ItemCardProps} from "@ecommerce/shared/ui/ItemCard";
import {NavButtonProps} from "@ecommerce/shared/ui/NavButton";
import {ProductGalleryProps} from "@ecommerce/shared/ui/ProductGallery";

import cat from './cat.jpg';
import cat2 from './cat2.jpeg';

console.log(cat2)

export const PRODUCT: ProductSidebarProps['product'] = {
  slug: 'some-name',
  price: 65,
  name: 'Some name',
  images: {
    data: [
      {
        attributes: {
          url: cat,
        },
      },
      {
        attributes: {
          url: cat2,
        },
      },
    ],
  },
  smallDescription: 'Small Description',
  description: 'Description',
  information: [
    {
      title: 'Title1',
      content: 'content1',
    },
  ],
};

export const ITEM: ItemCardProps['item'] = {
  slug: 'cat-cat',
  price: 65,
  name: 'cat cat',
  images: {
    data: [
      {
        attributes: {
          url: cat,
        },
      },
      {
        attributes: {
          url: cat2,
        },
      },
    ],
  },
  smallDescription: 'smallDescription',
  description: 'description',
};

export const BUTTON_DEFAULT_PROPS: ButtonProps = {
  children: 'hello',
  tag: 'button',
  size: 'base',
  type: 'primary',
  round: 'full',
};

export const CART_ITEM_DEFAULT_PROPS: CartItemProps = {
  inCartAmount: 1,
  item: PRODUCT,
  handleCartMinusClick: () => {
    console.log('cart minus');
  },
  handleCartPlusClick: () => {
    console.log('cart plus');
  },
  handleRemoveFromCart: () => {
    console.log('cart remove');
  },
  toggleFavorite: () => {
    console.log('favorite');
  },
};

export const CART_SIDEBAR_DEFAULT_PROPS: CartSidebarProps = {
  cartTotal: 100,
  isUser: true,
  handleCheckoutClick: async () => {
    console.log('checkout');
  },
  isLoading: false,
};

export const CATEGORY_CARD_DEFAULT_PROPS: CategoryCardProps = {
  category: {
    smallDescription:
      'smalll des',
    image: {
      data: {
        attributes: {
          url: cat,
        },
      },
    },
    slug: '1',
  },
};
export const ICON_DEFAULT_PROPS: IconProps = {
  icon: 'cart',
};

export const ITEM_CARD_DEFAULT_PROPS: ItemCardProps = {
  item: ITEM,
  size: 'base',
  handleCartClick: () => console.log('cart clicked'),
  handleAmountCartClick: () => console.log('amount cart clicked'),
  handleFavoriteClick: () => console.log('favorite clicked'),
};
export const NAV_BUTTON_DEFAULT_PROPS: NavButtonProps = {
  type: 'heart',
};
export const PRODUCT_GALLERY_DEFAULT_PROPS: ProductGalleryProps = {
  images: [
    {
      attributes: {
        url: cat,
      },
    },
    {
      attributes: {
        url: cat2,
      },
    },
  ],
};
