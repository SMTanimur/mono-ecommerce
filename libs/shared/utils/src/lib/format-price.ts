export const formatPrice = (price: number) =>
  `${Math.max(price, 0).toFixed(2)}$`;
