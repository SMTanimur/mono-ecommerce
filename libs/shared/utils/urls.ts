export const getApiUrl = (path = ''): string =>
  `${process.env['NX_STRAPI_IMG_URL'] ?? ''}${path}`;

export const getImageUrl = (path = ''): string =>
  `${process.env['NX_STRAPI_URL'] ?? 'http://localhost:4200'}${path}`;

export const createItemHref = (id: string): string => 
    `/i/${id}`;
