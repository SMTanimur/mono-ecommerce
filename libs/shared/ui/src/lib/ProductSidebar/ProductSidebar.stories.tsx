import { Story, Meta } from '@storybook/react';
import { PRODUCT_SIDEBAR_DEFAULT_PROPS } from '@ecommerce/shared/props/constants';
import { ProductSidebar, ProductSidebarProps } from './ProductSidebar';

export default {
  component: ProductSidebar,
  title: 'Components/Product/Sidebar',
} as Meta;

const Template: Story<ProductSidebarProps> = (args) => (
  <ProductSidebar {...args} />
);

export const Primary = Template.bind({});
Primary.args = {
  ...PRODUCT_SIDEBAR_DEFAULT_PROPS,
};
