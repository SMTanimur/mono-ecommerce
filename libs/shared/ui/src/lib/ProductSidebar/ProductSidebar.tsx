import { ProductSlugQuery } from '@ecommerce/shared/graphql/types';
import { formatPrice } from '@ecommerce/shared/utils/format-price';
import clsx from 'clsx';
import { Cart } from './Components/Cart';
import { Favorite } from './Components/Favorite';

export interface ProductSidebarProps {
  product: NonNullable<
    NonNullable<ProductSlugQuery['products']>['data'][0]['attributes']
    >;
  handleCartClick: () => void;
  handleRemoveCartClick: () => void;
  handleFavoriteClick: () => void;
  isFavorite: boolean;
  inCartAmount: number;
}

export const ProductSidebar: React.FC<ProductSidebarProps> = (props) => {
  const {
    product,
    handleCartClick,
    handleFavoriteClick,
    handleRemoveCartClick,
    inCartAmount,
    isFavorite
  } = props;

  const { name, price, smallDescription } = product;

  return (
    <aside
      className={clsx(
        'h-full min-w-56 w-full',
        'flex flex-col',
        'md:(order-2 top-12 sticky w-2/5)'
      )}
    >
      <div className={clsx('mb-2', 'flex items-center justify-between')}>
        <h1
          className={clsx(
            'max-w-3/5',
            'font-bold text-3xl uppercase break-words'
          )}
        >
          {name}
        </h1>
        <span className="font-semibold text-base">{formatPrice(price)}</span>
      </div>
      <p className={clsx('text-sm', 'opacity-75')}>{smallDescription}</p>
      <div className={clsx('mt-8', 'flex gap-4 justify-between')}>
        <Cart
          inCartAmount={inCartAmount}
          handleCartClick={handleCartClick}
          handleRemoveCartClick={handleRemoveCartClick}
        />
        <Favorite
          isFavorite={isFavorite}
          handleFavoriteClick={handleFavoriteClick}
        />
      </div>
    </aside>
  );
};
