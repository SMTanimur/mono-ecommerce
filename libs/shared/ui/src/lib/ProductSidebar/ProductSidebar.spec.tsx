import '@testing-library/jest-dom';
import { RenderResult, render, screen } from '@testing-library/react';
import { create } from 'react-test-renderer';

import { formatPrice } from '@ecommerce/shared/utils/format-price';

import { PRODUCT_SIDEBAR_DEFAULT_PROPS } from '../../props/constants';
import { ProductSidebar } from './ProductSidebar';

describe('ProductSidebar', () => {
    let renderResult: RenderResult;

    beforeEach(() => {
        renderResult = render(
            <ProductSidebar {...PRODUCT_SIDEBAR_DEFAULT_PROPS} />,
        );
    });

    it('should render successfully', () => {
        const { baseElement } = renderResult;

        const tree = create(
            <ProductSidebar {...PRODUCT_SIDEBAR_DEFAULT_PROPS} />,
        ).toJSON();
        expect(tree).toMatchSnapshot();

        expect(baseElement).toBeTruthy();
    });

    it('should have proper information', () => {
        const { smallDescription, name, price } =
            PRODUCT_SIDEBAR_DEFAULT_PROPS.product;

        expect(screen.getByText(smallDescription)).toBeInTheDocument();

        expect(screen.getByText(name)).toBeInTheDocument();

        expect(screen.queryByText(price)).not.toBeInTheDocument();
        expect(screen.queryByText(formatPrice(price))).toBeInTheDocument();
    });
});
