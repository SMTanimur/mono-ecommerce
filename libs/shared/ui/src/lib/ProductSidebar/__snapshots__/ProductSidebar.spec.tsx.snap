// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductSidebar should render successfully 1`] = `
<aside
  className="h-full min-w-56 w-full flex flex-col md:(order-2 top-12 sticky w-2/5)"
>
  <div
    className="mb-2 flex items-center justify-between"
  >
    <h1
      className="max-w-3/5 font-bold text-3xl uppercase break-words"
    >
      Some name
    </h1>
    <span
      className="font-semibold text-base"
    >
      65.00$
    </span>
  </div>
  <p
    className="text-sm opacity-75"
  >
    Small Description
  </p>
  <div
    className="mt-8 flex gap-4 justify-between"
  >
    <div
      className="flex-1 relative"
    >
      <button
        className="inline-block cursor-pointer transform-gpu transition-all active:scale-92 rounded-full w-full h-full text-white bg-sky-600 hover:(bg-sky-700) active:(bg-sky-800)"
        onClick={[Function]}
      >
        <span
          className="font-semibold"
        >
          Add to bag
        </span>
      </button>
    </div>
    <button
      className="inline-block cursor-pointer transform-gpu transition-all active:scale-92 rounded-full p-4 hover:(bg-gray-100 border-black) active:(bg-gray-300) border border-gray-300 flex justify-center items-center"
      onClick={[Function]}
      title="add to favorite"
      type="button"
    >
      <div
        aria-hidden="true"
        className="relative flex items-center select-none pointer-events-none h-4 w-4  sm:(h-6 w-6)"
      >
        <svg
          className="feather feather-heart "
          fill="red"
          height={24}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          viewBox="0 0 24 24"
          width={24}
        >
          <g
            dangerouslySetInnerHTML={
              Object {
                "__html": "<path d=\\"M20.84 4.61a5.5 5.5 0 0 0-7.78 0L12 5.67l-1.06-1.06a5.5 5.5 0 0 0-7.78 7.78l1.06 1.06L12 21.23l7.78-7.78 1.06-1.06a5.5 5.5 0 0 0 0-7.78z\\"></path>",
              }
            }
          />
        </svg>
      </div>
    </button>
  </div>
</aside>
`;
